gateway/tests/gateway_test.rs

```diff
@@ -675,40 +680,300 @@ fn test_send_cross() {
+    // Part 1: test the message is stored in postbox
+    let st: State = rt.get_state();
+    assert_ne!(tt.subnet().unwrap(), st.network_name);
+
+    // Check 1: `tt` is in `sub1`, which is not in that of `runtime` of gateway, will store in postbox
+    let item = st.load_from_postbox(rt.store(), cid.clone()).unwrap();
+    assert_eq!(item.owners, Some(vec![ff.clone().raw_addr().unwrap()]));
+    let msg = item.cross_msg.msg;
+    assert_eq!(msg.to, tt);
+    // the nonce should not have changed at all
+    assert_eq!(msg.nonce, msg_nonce);
+    assert_eq!(msg.value, value);
+
+    // Part 2: Now we propagate from postbox
+    // get the original subnet nonce first
+    let caller = ff.clone().raw_addr().unwrap();
+    h.propagate(&mut rt, caller, cid.clone()).unwrap();
+
+    // state should be updated, load again
+    let st: State = rt.get_state();
+
+    // cid should be removed from postbox
+    let r = st.load_from_postbox(rt.store(), cid.clone());
+    assert_eq!(r.is_err(), true);
+    let err = r.unwrap_err();
+    assert_eq!(err.to_string(), "cid not found in postbox");
+}
+
+/// This test covers the case where a bottom up cross_msg's target subnet is NOT the same as that of
+/// the gateway. It would save in postbox. Also, the gateway is the nearest parent, a switch to
+/// top down cross msg should occur.
+#[test]
+fn test_apply_msg_bu_switch_td() {
+    // ============== Register subnet ==============
+    let parent_sub = SubnetID::new_from_parent(&ROOTNET_ID, *SUBNET_ONE);
+    let (h, mut rt) = setup(parent_sub.clone());
+
+    let from = Address::new_bls(&[3; fvm_shared::address::BLS_PUB_LEN]).unwrap();
+    let to = Address::new_bls(&[4; fvm_shared::address::BLS_PUB_LEN]).unwrap();
+
+    // ================ Setup ===============
+    let value = TokenAmount::from_atto(10_u64.pow(17));
+
+    // ================= Bottom-Up ===============
+    let reg_value = TokenAmount::from_atto(10_u64.pow(18));
+    // ff: /root/f101/f102
+    // to: /root/f101/f103
+    // we are executing the message from, harness or the gateway is at: /root/f101
+    let ff_sub = SubnetID::new_from_parent(&parent_sub, *SUBNET_TWO);
+    let tt_sub = SubnetID::new_from_parent(&parent_sub, *SUBNET_THR);
     h.register(&mut rt, &SUBNET_TWO, &reg_value, ExitCode::OK)
         .unwrap();
+    h.register(&mut rt, &SUBNET_THR, &reg_value, ExitCode::OK)
+        .unwrap();
+
+    let ff = IPCAddress::new(&ff_sub, &to).unwrap();
+    let tt = IPCAddress::new(&tt_sub, &from).unwrap();
+    let msg_nonce = 0;
+
+    // Only system code is allowed to this method
+    let params = StorableMsg {
+        to: tt.clone(),
+        from: ff.clone(),
+        method: METHOD_SEND,
+        value: value.clone(),
+        params: RawBytes::default(),
+        nonce: msg_nonce,
+    };
+
+    let caller = ff.clone().raw_addr().unwrap();
+
+    // we directly insert message into postbox as we dont really care how it's got stored in queue
+    let mut cid = Cid::default();
```

Comment:
wait, wait, wait... if `apply_msg` is implemented correctly, you should be able to call `h.apply_cross_execute_only(` as you do in the rest of the tests to introduce the message in the postbox. If this is not the case it means that we have something wrong. Can you change it? Thanks!

Reply:
yeah, should just return `cid` from transaction directly, I updated.